### ISSUE TITLE
Automatically determine address

### ISSUE THREAD

**Issue by RayCarrot** [mateusfavarin/psx-modding-toolchain#6]
I'm unsure how reasonable this would be, but I was thinking if there could be a way to automatically determine the address to compile .c files for (based on user-defined memory ranges)?

Basically I have several smaller mods for a game, and compiling them each one by one works fine. However the issue comes when I want to combine multiple of them. Now if these mods don't share any of the same addresses for the .c files then it works fine, but it's a lot of work manually managing that. It means constantly checking the size of the output and seeing where in the exe there's enough of unused space to place that specific mod.

What I would propose would be in the `buildlist` to be able to set the address to `auto` and then somewhere else, like the config, you can specify available memory ranges which can be used. In my case this would consists of a lot of memory ranges in the exe where there are unused functions (mostly unused Psyq functions) which mods can freely overwrite. Then there'd be an option to build multiple mods at once where it'd make sure each mod uses a unique address from the range that fits the size of that mod.

**Comment by mateusfavarin** [comment]
I agree that dealing with multiple mods as of now is cumbersome. I like the idea of having an `auto` memory range that could be filled with a smart heuristic, but that'd require changing the inner workings of the compilation system so I'm labing this issue as `long term`. I will be working on achieving this end goal by dividing this issue into sub-tasks and implementing them:

- [x] Support adding multiple mods to the buildList, so that the program triggers chain-compilations for each mod added in different contexts, allowing the inclusion of different programs in one final output. Implemented via [1fd5006](https://github.com/mateusfavarin/psx-modding-toolchain/commit/1fd5006c4473c05111cd49f68ee94a3b9079fe3e). You can now combine mods by adding the line `add, ModFolder` to the build list.
- [ ] Divide the compilation process into compiling and linking, so that every mod can first be compiled, then the memory used by every section of each mod can be calculated, and finally the results can be used in a memory range heuristic in order to determine where each section should belong to.
- [ ] Change the way output files are handled by writing down a list of each valid compiled file and its final address, removing a second unnecessary parse of the build list and having a place to look up where each section was rellocated.
